### PR TITLE
209: Bug fixes

### PR DESCRIPTION
**Pull request by qdddddd**

### CODE REVIEW

File: src/epic-cli.cpp [hunk 1]
```diff
@@ -139,7 +139,6 @@ int main(int argc, char** argv) {
         auto client = CreateClient("0.0.0.0", rpc_port);
 
         switch (commandMap.at(command)) {
-
```
Comment (0ersted) [inline src/epic-cli.cpp]: this line is added by design

File: src/messages/block.cpp [hunk 2]
```diff
@@ -58,14 +58,22 @@ Block::Block() : NetMessage(BLOCK) {
...
     SetParents();
 };
 
 Block::Block(Block&& b) noexcept
     : NetMessage(BLOCK), hash_(b.hash_), header_(b.header_), proof_(std::move(b.proof_)),
-      transactions_(std::move(b.transactions_)), optimalEncodingSize_(b.optimalEncodingSize_), source(b.source) {
+      optimalEncodingSize_(b.optimalEncodingSize_), source(b.source) {
+    for (auto& ptx : b.transactions_) {
+        transactions_.emplace_back(std::make_shared<Transaction>(std::move(*ptx)));
+    }
```
Comment (0ersted) [inline src/messages/block.cpp]: I think in the move ctor we can just move the vector as the old block will be set null

File: src/rpc/rpc_client.cpp [hunk 3]
```diff
@@ -251,7 +252,7 @@ RPCClient::option_string RPCClient::Login(const std::string& passphrase) {
     grpc::ClientContext context;
 
     request.set_passphrase(passphrase);
-   auto status = commander_stub_->Login(&context, request, &response);
+    auto status = commander_stub_->Login(&context, request, &response);
```
Comment (0ersted) [inline src/rpc/rpc_client.cpp]: ...

File: src/rpc/rpc_server.cpp [hunk 4]
```diff
@@ -121,8 +121,11 @@ grpc::Status CommanderRPCServiceImpl::CreateRandomTx(grpc::ServerContext* contex
     if (!WALLET) {
         reply->set_result("Wallet has not been started");
     } else {
-        WALLET->CreateRandomTx(request->size());
-        reply->set_result("Now wallet is creating tx");
+        if (WALLET->CreateRandomTx(request->size())) {
+            reply->set_result("Now wallet is creating tx");
+        } else {
+            reply->set_result("Wallet is not ready. Did you forget to set passphrase?");
+        }
```
Comment (0ersted) [inline src/rpc/rpc_server.cpp]: I supplement this logic in #137  so I recommend removing it here

File: src/wallet/wallet.cpp [hunk 5]
```diff
@@ -332,7 +332,11 @@ bool Wallet::CanRedeem(Coin coins) {
     return GetMinerInfo().second >= coins && pendingRedemption.empty();
 }
 
-void Wallet::CreateRandomTx(size_t sizeTx) {
+bool Wallet::CreateRandomTx(size_t sizeTx) {
+    if (!crypter_.IsReady()) {
+        return false;
+    }
+
```
Comment (0ersted) [inline src/wallet/wallet.cpp]: same as above

File: src/epic-cli.cpp (hunk 1 above)
Comment (qdddddd) [reply src/epic-cli.cpp]: Removed by clang-format. Add `// clang-format off` ?

File: src/rpc/rpc_client.cpp (hunk 3 above)
Comment (qdddddd) [reply src/rpc/rpc_client.cpp]: The power of clang-format!

File: src/messages/block.cpp (hunk 2 above)
Comment (qdddddd) [reply src/messages/block.cpp]: Works

File: src/rpc/rpc_server.cpp (hunk 4 above)
Comment (qdddddd) [reply src/rpc/rpc_server.cpp]: We can merge it. There’s also changes in Wallet

File: config.toml [hunk 6]
```diff
@@ -5,7 +5,7 @@ root = "data/"
 
 [logs]
 level = "info"
-use_file_logger = true
+use_file_logger = false
```
[the remark this answers is not in the document]
Comment (qdddddd) [reply config.toml]: Now it is not 👀

File: src/pow/miner.cpp [hunk 7]
```diff
@@ -203,23 +203,25 @@ void Miner::Run() {
...
-                        b.AddTransaction(std::move(tx));
-                    }
-
                     auto allowed =
                         CalculateAllowedDist(distanceCal_, std::atomic_load(&chainHead_)->snapshot->hashRate);
                     b.AddTransactions(MEMPOOL->ExtractTransactions(prevHash, allowed, GetParams().blockCapacity));
-                    b.SetMerkle();
                 }
+
+                b.SetMerkle();
```
Comment (bassland) [inline src/pow/miner.cpp]: I suggest combine `SetMerkle` line 203 and line 224 into one line before `Solve`

**Pull request merged**